packages/truffle-compile/compilerSupplier/index.js

```diff
@@ -0,0 +1,96 @@
+const path = require("path");
+const fs = require("fs");
+const semver = require("semver");
+
+const {
+  Bundled,
+  Docker,
+  LoadingStrategy,
+  Local,
+  Native,
+  VersionRange
+} = require("./loadingStrategies");
+
+class CompilerSupplier {
+  constructor(_config) {
+    _config = _config || {};
+    const defaultConfig = { version: null };
+    this.config = Object.assign({}, defaultConfig, _config);
+    this.strategyOptions = { version: this.config.version };
+  }
+
+  badInputError(userSpecification) {
+    const message =
+      `Could not find a compiler version matching ${userSpecification}. ` +
+      `compilers.solc.version option must be a string specifying:\n` +
+      `   - a path to a locally installed solcjs\n` +
+      `   - a solc version or range (ex: '0.4.22' or '^0.5.0')\n` +
+      `   - a docker image name (ex: 'stable')\n` +
+      `   - 'native' to use natively installed solc\n`;
+    return new Error(message);
+  }
+
+  load() {
+    const userSpecification = this.config.version;
+
+    return new Promise(async (resolve, reject) => {
+      let strategy;
+      const useDocker = this.config.docker;
+      const useNative = userSpecification === "native";
+      const useBundledSolc = !userSpecification;
+      const useSpecifiedLocal =
+        userSpecification && this.fileExists(userSpecification);
+      const isValidVersionRange = semver.validRange(userSpecification);
+
+      if (useDocker) {
+        strategy = new Docker(this.strategyOptions);
+        return resolve(await strategy.load(userSpecification));
+      }
+      if (useNative) {
+        strategy = new Native(this.strategyOptions);
+        return resolve(await strategy.load(userSpecification));
+      }
+      if (useBundledSolc) {
+        strategy = new Bundled(this.strategyOptions);
+        return resolve(await strategy.load(userSpecification));
+      }
+      if (useSpecifiedLocal) {
+        strategy = new Local(this.strategyOptions);
+        return resolve(await strategy.load(userSpecification));
+      }
+      if (isValidVersionRange) {
+        strategy = new VersionRange(this.strategyOptions);
+        return resolve(await strategy.load(userSpecification));
```

Comment:
We can move `strategy.load()` out of the block of `if` statements.

Reply:
Hmm, how about a method that returns the strategy and...
```
if (strategy) {
  strategy.load(userSpecification);
} else {
  throw Error....
```?

Reply:
Doesn't need to be a separate method, but yeah, just at the end have that `if(strategy)` check + load.